### PR TITLE
[FLINK-12970][hive] Support writing Hive complex types

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

To support writing to Hive tables with complex data types.

## Brief change log

  - Support writing ARRAY, MAP and STRUCT types in `HiveTableOutputFormat`.
  - Use `HiveObjectConversion` to convert Flink object to Hive object in `HiveTableOutputFormat`.
  - Fixed some issue with creating Hive TypeInfo for struct type.

## Verifying this change

Added new unit test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by lirui-apache** [comment]
cc @xuefuz @bowenli86 @zjuwangg

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java [hunk 1]
```diff
@@ -146,15 +150,31 @@ public static ObjectInspector getObjectInspector(DataType flinkType) throws IOEx
 		return getObjectInspector(HiveTypeUtil.toHiveTypeInfo(flinkType));
 	}
 
-	// TODO: reuse Hive's TypeInfoUtils?
 	private static ObjectInspector getObjectInspector(TypeInfo type) throws IOException {
```
Comment (bowenli86) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/util/HiveTableUtil.java]: `HiveInspectors` also has such a method. Can we move the whole thing to `HiveInspectors` so we have a single place to do all kinds of conversions?

**Comment by lirui-apache** [comment]
Thanks @bowenli86 for the review. Please take another look.

**Comment by bowenli86** [comment]
LGTM. Merging

**Pull request closed without merging**